### DISCUSSION TITLE
using Google collab

### DISCUSSION THREAD

**Discussion by tataphani** [sassoftware/saspy#439] [General]
hello all, 

I think you got this questions earlier too, but I could not find right answer . I am SAS user and I want to learn Python . So  saspy is answer . But to use it  where and how I can use it . I do not have access to SAS on local machine , student edition is out as of 02/03/2022 . when working on my office laptop ( behind VPN ) google collab seems to solution for my need . Can you walk me through the steps please . Working on a presentation to show my collogues about the SASPy

**Comment by tomweber-sas** [comment]
Hey @tataphani, so where is SAS that you are going to try to connect to, and what kind of deployment is it? See [here](https://sassoftware.github.io/saspy/configuration.html#choosing-an-access-method) to know which saspy access method can be used depending upon the SAS deployment and the client location. You mention student edition, was that University Edition; a VM on your local machine? The replacement for University Edition, is SAS On Demand for Academics (SAS/ODA). If that's what you are going to connect to, the configuration instructions are [here](https://support.sas.com/ondemand/saspy.html). If you're connecting to a SAS sever in your office, then how you configure that depends on how it's deployed; stand alone, or as a workspace server. So, knowing what kind of SAS server you're connecting to, and where it's located, is the first thing to understand.

**Comment by tataphani** [comment]
Thanks a lot for quick answer , I am not allowed to SAS for this personnel presentation . SO I am using SAS/ODA and google Collab to connect to it .  The instructions given at this page https://support.sas.com/ondemand/saspy.html are also true for connecting it to  Collab ? reading on the web 

[How to run SAS in Google Colab]()
The above step by step instructions are mainly designed for running python in Jupyter notebook which is the most commonly used interface for Python. Recently Google Colab has become a go-to tool for data science because of several reasons - supports version controlling, notebooks saved in Google Drive, work from anywhere, supports GPU etc. In simple words it runs on cloud so you don't need to install python and popular python packages. Sharing code with your coworkers is also very easy and effective via colab. Java is already installed on colab. You just need to specify this file location /usr/bin/java for java in step 2 (listed above).

import saspy
sas = saspy.SASsession(java='/usr/bin/java', iomhost=['odaws01-apse1.oda.sas.com','odaws02-apse1.oda.sas.com'], iomport=8591, encoding='utf-8')
sas
Make sure to check iomhost as per your region. Read Step 2 above.
#US Home Region
iomhost = ['odaws01-usw2.oda.sas.com','odaws02-usw2.oda.sas.com','odaws03-usw2.oda.sas.com','odaws04-usw2.oda.sas.com']


is this enough ? 


%%SAS sas magic does not work so you can use sas.submitLST( ) like below.
sas.submitLST("proc print data=sashelp.cars; run;", method='listorlog')

**Reply by tomweber-sas** [reply]
I just went to Colab and made a connection to ODA as you show above:

```
# needed to install saspy and see where java was installed first
! pip install saspy
! which java

# then I was able to connect
import saspy
sas = saspy.SASsession(iomhost=['odaws01-usw2.oda.sas.com', 'odaws02-usw2.oda.sas.com',
 'odaws03-usw2.oda.sas.com', 'odaws04-usw2.oda.sas.com'],
 java='/usr/bin/java', iomport=8591)

Using SAS Config named: default
Please enter the IOM user id: xxxxx
Please enter the password for IOM user : ··········
SAS Connection established. Subprocess id is 545

Access Method         = IOM
SAS Config name       = default
SAS Config file       = /usr/local/lib/python3.7/dist-packages/saspy/sascfg.py
WORK Path             = /saswork/SAS_workF3F50000E185_odaws01-usw2.oda.sas.com/SAS_work0D7F0000E185_odaws01-usw2.oda.sas.com/
SAS Version           = 9.04.01M6P11072018
SASPy Version         = 4.0.0
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = utf-8
Python Encoding value = utf_8
SAS process Pid value = 57733
```

Were you able to get it to connect too? Seems pretty easy.

![image](https://user-images.githubusercontent.com/17710182/152385600-057b8a09-f5de-4ebe-861d-f8bac9ac6849.png)

**Reply by tataphani** [reply]
Thanks a lot for trying it out , will try it out in the afternoon today , very thankful for your support

**Reply by tomweber-sas** [reply]
cool, let me know how it goes!

**Comment by tataphani** [comment]
Does SAS session has to be logged in always ?

**Comment by tomweber-sas** [comment]
As far as the SAS server (Workspace server) you're connecting to, I believe it has an inactivity timeout setting at 3 hours. It will disconnect after not being used. Then you would need to log in to a new server. There may be other constraints on the ODA system too, that I'm not familiar with. How long the Colab session stays active is another thing, and I don't know about that. But if your Python session times out and you have to restart it (or you restart it yourself), then, of course, your connection to a server is gone and you would have to establish a new one too.
Not completely sure if this answers your question.
